### PR TITLE
PP-186: Refactor bookmark service

### PR DESCRIPTION
**Pull request by io7m**
**What's this do?**
This creates a new "create bookmark" operation in the bookmark service that will first create a local bookmark, and then try to create an equivalent remote bookmark. The operation can be configured to ignore remote failures. This refactors the existing operations to return either non-nullable Bookmarks, or fail the resulting Future values with exceptions.

This also refactors the "delete bookmark" operation similarly. The deletion occurs in reverse order; a remote bookmark is deleted first, and then the corresponding local bookmark is deleted if the remote bookmark deletion succeeds. This is also, currently, configured to ignore remote failures at the current call sites.

These changes allow for having the bookmark service do the heavy lifting, and simplifies the various viewers that use it.

**Why are we doing this? (w/ JIRA link if applicable)**
Affects: https://ebce-lyrasis.atlassian.net/browse/PP-186

**How should this be tested? / Do these changes have associated tests?**
Try creating and deleting bookmarks in the audio book player and epub viewer, both with and without annotation syncing enabled.

**Dependencies for merging? Releasing to production?**
None.

**Have you updated the changelog?**
It was updated in the previous bookmark PR.

**Has the application documentation been updated for these changes?**
Only comments.

**Did someone actually run this code to verify it works?**
@io7m tried various audio books and epubs.

### CODE REVIEW

File: simplified-viewer-audiobook/src/main/java/org/nypl/simplified/viewer/audiobook/AudioBookPlayerActivity.kt [hunk 1]
```diff
@@ -971,22 +971,18 @@ class AudioBookPlayerActivity :
...
-        bookmark = bookmark
-      )
-
-      if (wasDeleted) {
-        currentBookmarks.remove(bookmark)
+    this.bookmarkService.bookmarkDelete(
+      accountID = this.parameters.accountID,
+      bookmark = bookmark,
+      ignoreRemoteFailures = true
+    ).map { wasDeleted ->
```
Comment (nunommts) [inline simplified-viewer-audiobook/src/main/java/org/nypl/simplified/viewer/audiobook/AudioBookPlayerActivity.kt:978]: This `wasDeleted` variable doesn't need to exist now the method is a Unit and doesn't return a boolean, right?

File: simplified-viewer-epub-readium2/src/main/java/org/nypl/simplified/viewer/epub/readium2/Reader2Activity.kt [hunk 2]
```diff
@@ -463,11 +465,13 @@ class Reader2Activity : AppCompatActivity(R.layout.reader2) {
...
         bookmarkService.bookmarkDelete(
           accountID = account.id,
-          bookmark = bookmark
-        ).onAnyError { false }
-          .map { wasDeleted ->
-            event.onDeleteOperationFinished(wasDeleted)
-          }
+          bookmark = bookmark,
+          ignoreRemoteFailures = true
+        ).map { wasDeleted ->
```
Comment (nunommts) [inline simplified-viewer-epub-readium2/src/main/java/org/nypl/simplified/viewer/epub/readium2/Reader2Activity.kt:470]: Same here!

**Review by nunommts — CHANGES_REQUESTED** [review]
This looks good! I just have two minor comments/suggestions and a more "overall" question:

1 - Is there any reason to have the `ignoreRemoteFailures`variable since it's always true?

**Comment by io7m** [comment]
> This looks good! I just have two minor comments/suggestions and a more "overall" question:
> 
> 1 - Is there any reason to have the `ignoreRemoteFailures`variable since it's always true?

I'm not sure there's a good answer. I was trying to avoid changing the policy too much, but combining the local and remote creation of bookmarks into a single service method meant that it would no longer be possible to distinguish between either one failing; this would mean that in the pretty common case of a remote bookmark failing, no local bookmark would appear to have been created. I thought it might have been a step too far to have the code unconditionally ignore any and every remote failure, so I left it parameterized.

My guess is that we'll be revisiting all of this soon, given that the default is changing and we have no behaviour spec with regards to how the apps should behave when things fail... :sweat:

File: simplified-viewer-audiobook/src/main/java/org/nypl/simplified/viewer/audiobook/AudioBookPlayerActivity.kt (hunk 1 above)
Comment (io7m) [reply simplified-viewer-audiobook/src/main/java/org/nypl/simplified/viewer/audiobook/AudioBookPlayerActivity.kt]: Oops!

**Comment by nunommts** [comment]
@io7m ok, got it. Since it's fairly easy to remove in the future I don't see any harm on keeping it as it is right now, specially if you're planning to revisit this soon.

**Review by nunommts — APPROVED** [review]

**Pull request merged**